File: tests/wasm/test_random_walk_index.js
import test from 'node:test';
import assert from 'node:assert';
import path from 'path';
import { fileURLToPath } from 'url';
import { loadTestData, assertArrayClose } from './test_utils.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = path.dirname(__filename);

let wasm;
let testData;

function manualRandomWalkIndex(high, low, close, length) {
    const n = close.length;
    const outHigh = new Float64Array(n);
    const outLow = new Float64Array(n);
    outHigh.fill(NaN);
    outLow.fill(NaN);

    let first = -1;
    for (let i = 0; i < n; i++) {
        if (Number.isFinite(high[i]) && Number.isFinite(low[i]) && Number.isFinite(close[i])) {
            first = i;
            break;
        }
    }
    if (first === -1) {
        return { high: outHigh, low: outLow };
    }

    const warm = first + length - 1;
    const alpha = 1 / length;
    const sqrtLength = Math.sqrt(length);
    let prevClose = close[first];
    let sumTr = high[first] - low[first];
    let atr = NaN;

    if (length === 1) {
        atr = sumTr;
        const denom = atr * sqrtLength;
        if (Number.isFinite(denom) && denom !== 0) {
            outHigh[first] = high[first] / denom;
            outLow[first] = -low[first] / denom;
        }
    }

    for (let i = first + 1; i < n; i++) {
        const tr = Math.max(
            high[i] - low[i],
            Math.abs(high[i] - prevClose),
            Math.abs(low[i] - prevClose),
        );
        if (i <= warm) {
            sumTr += tr;
            if (i === warm) {
                atr = sumTr / length;
            }
        } else {
            atr = atr + alpha * (tr - atr);
        }

        if (i >= warm) {
            const histLow = i >= length && Number.isFinite(low[i - length]) ? low[i - length] : 0;
            const histHigh = i >= length && Number.isFinite(high[i - length]) ? high[i - length] : 0;
            const denom = atr * sqrtLength;
            if (Number.isFinite(denom) && denom !== 0) {
                outHigh[i] = (high[i] - histLow) / denom;
                outLow[i] = (histHigh - low[i]) / denom;
            }
        }
        prevClose = close[i];
    }

    return { high: outHigh, low: outLow };
}

test.before(async () => {
    const wasmPath = path.join(__dirname, '../../pkg/vector_ta.js');
    const importPath = process.platform === 'win32'
        ? 'file:///' + wasmPath.replace(/\\/g, '/')
        : wasmPath;
    wasm = await import(importPath);
    testData = loadTestData();
});

test('random_walk_index_js matches manual reference', () => {
    const n = 160;
    const high = new Float64Array(testData.high.slice(0, n));
    const low = new Float64Array(testData.low.slice(0, n));
    const close = new Float64Array(testData.close.slice(0, n));
    const result = wasm.random_walk_index_js(high, low, close, 14);
    const expected = manualRandomWalkIndex(high, low, close, 14);
    assertArrayClose(result.high, expected.high, 1e-12, 'high mismatch');
    assertArrayClose(result.low, expected.low, 1e-12, 'low mismatch');
});

test('random_walk_index_batch_js first row matches single', () => {
    const n = 96;
    const high = new Float64Array(testData.high.slice(0, n));
    const low = new Float64Array(testData.low.slice(0, n));
    const close = new Float64Array(testData.close.slice(0, n));
    const batch = wasm.random_walk_index_batch_js(high, low, close, {
        length_range: [14, 16, 2],
    });
    const single = wasm.random_walk_index_js(high, low, close, 14);
    assert.strictEqual(batch.rows, 2);
    assert.strictEqual(batch.cols, n);
    assert.deepStrictEqual(batch.lengths, [14, 16]);
    assertArrayClose(batch.high.slice(0, n), single.high, 1e-12, 'batch high mismatch');
    assertArrayClose(batch.low.slice(0, n), single.low, 1e-12, 'batch low mismatch');
});

test('random_walk_index_into pointer path matches safe API', () => {
    const n = 128;
    const high = new Float64Array(testData.high.slice(0, n));
    const low = new Float64Array(testData.low.slice(0, n));
    const close = new Float64Array(testData.close.slice(0, n));
    const safe = wasm.random_walk_index_js(high, low, close, 14);

    const highPtr = wasm.allocate_f64_array(n);
    const lowPtr = wasm.allocate_f64_array(n);
    const closePtr = wasm.allocate_f64_array(n);
    const outHighPtr = wasm.random_walk_index_alloc(n);
    const outLowPtr = wasm.random_walk_index_alloc(n);

    try {
        wasm.write_f64_array(highPtr, high);
        wasm.write_f64_array(lowPtr, low);
        wasm.write_f64_array(closePtr, close);
        wasm.random_walk_index_into(highPtr, lowPtr, closePtr, outHighPtr, outLowPtr, n, 14);
        const gotHigh = wasm.read_f64_array(outHighPtr, n);
        const gotLow = wasm.read_f64_array(outLowPtr, n);
        assertArrayClose(gotHigh, safe.high, 1e-12, 'pointer high mismatch');
        assertArrayClose(gotLow, safe.low, 1e-12, 'pointer low mismatch');
    } finally {
        wasm.random_walk_index_free(outHighPtr, n);
        wasm.random_walk_index_free(outLowPtr, n);
        wasm.deallocate_f64_array(highPtr);
        wasm.deallocate_f64_array(lowPtr);
        wasm.deallocate_f64_array(closePtr);
    }
});

test('random_walk_index_batch_into returns row count', () => {
    const n = 96;
    const high = new Float64Array(testData.high.slice(0, n));
    const low = new Float64Array(testData.low.slice(0, n));
    const close = new Float64Array(testData.close.slice(0, n));
    const highPtr = wasm.allocate_f64_array(n);
    const lowPtr = wasm.allocate_f64_array(n);
    const closePtr = wasm.allocate_f64_array(n);
    const rows = 2;
    const outHighPtr = wasm.random_walk_index_alloc(rows * n);
    const outLowPtr = wasm.random_walk_index_alloc(rows * n);

    try {
        wasm.write_f64_array(highPtr, high);
        wasm.write_f64_array(lowPtr, low);
        wasm.write_f64_array(closePtr, close);
        const returnedRows = wasm.random_walk_index_batch_into(
            highPtr,
            lowPtr,
            closePtr,
            outHighPtr,
            outLowPtr,
            n,
            14,
            16,
            2,
        );
        assert.strictEqual(returnedRows, rows);
    } finally {
        wasm.random_walk_index_free(outHighPtr, rows * n);
        wasm.random_walk_index_free(outLowPtr, rows * n);
        wasm.deallocate_f64_array(highPtr);
        wasm.deallocate_f64_array(lowPtr);
        wasm.deallocate_f64_array(closePtr);
    }
});

test('random_walk_index_js rejects invalid length', () => {
    const n = 64;
    const high = new Float64Array(testData.high.slice(0, n));
    const low = new Float64Array(testData.low.slice(0, n));
    const close = new Float64Array(testData.close.slice(0, n));
    assert.throws(() => {
        wasm.random_walk_index_js(high, low, close, 0);
    }, /Invalid length/);
});

File: tests/python/test_random_walk_index.py
import numpy as np
import pytest

from test_utils import load_test_data

try:
    import vector_ta as mp
except ImportError:
    pytest.skip(
        "Python module not built. Run 'maturin develop --features python' first",
        allow_module_level=True,
    )


def manual_random_walk_index(high: np.ndarray, low: np.ndarray, close: np.ndarray, length: int):
    n = close.shape[0]
    out_high = np.full(n, np.nan, dtype=np.float64)
    out_low = np.full(n, np.nan, dtype=np.float64)

    valid = np.flatnonzero(np.isfinite(high) & np.isfinite(low) & np.isfinite(close))
    if valid.size == 0:
        return out_high, out_low

    first = int(valid[0])
    warm = first + length - 1
    alpha = 1.0 / float(length)
    sqrt_length = np.sqrt(float(length))

    prev_close = close[first]
    sum_tr = high[first] - low[first]
    atr = np.nan

    if length == 1:
        atr = sum_tr
        denom = atr * sqrt_length
        if np.isfinite(denom) and denom != 0.0:
            out_high[first] = high[first] / denom
            out_low[first] = -low[first] / denom

    for i in range(first + 1, n):
        tr = max(
            high[i] - low[i],
            abs(high[i] - prev_close),
            abs(low[i] - prev_close),
        )
        if i <= warm:
            sum_tr += tr
            if i == warm:
                atr = sum_tr / float(length)
        else:
            atr = atr + alpha * (tr - atr)

        if i >= warm:
            hist_low = low[i - length] if i >= length and np.isfinite(low[i - length]) else 0.0
            hist_high = high[i - length] if i >= length and np.isfinite(high[i - length]) else 0.0
            denom = atr * sqrt_length
            if np.isfinite(denom) and denom != 0.0:
                out_high[i] = (high[i] - hist_low) / denom
                out_low[i] = (hist_high - low[i]) / denom

        prev_close = close[i]

    return out_high, out_low


class TestRandomWalkIndex:
    def test_manual_reference_matches_binding(self):
        data = load_test_data()
        high = np.asarray(data["high"][:160], dtype=np.float64)
        low = np.asarray(data["low"][:160], dtype=np.float64)
        close = np.asarray(data["close"][:160], dtype=np.float64)
        expected_high, expected_low = manual_random_walk_index(high, low, close, 14)
        result = mp.random_walk_index(high, low, close, length=14)
        np.testing.assert_allclose(result["high"], expected_high, rtol=0.0, atol=1e-12, equal_nan=True)
        np.testing.assert_allclose(result["low"], expected_low, rtol=0.0, atol=1e-12, equal_nan=True)

    def test_stream_matches_batch(self):
        data = load_test_data()
        high = np.asarray(data["high"][:128], dtype=np.float64)
        low = np.asarray(data["low"][:128], dtype=np.float64)
        close = np.asarray(data["close"][:128], dtype=np.float64)
        batch = mp.random_walk_index(high, low, close, length=14)
        stream = mp.RandomWalkIndexStream(14)
        streamed = np.array([stream.update(float(h), float(l), float(c)) for h, l, c in zip(high, low, close)])
        np.testing.assert_allclose(streamed[:, 0], batch["high"], rtol=0.0, atol=1e-12, equal_nan=True)
        np.testing.assert_allclose(streamed[:, 1], batch["low"], rtol=0.0, atol=1e-12, equal_nan=True)

    def test_batch_metadata_and_first_row_match_single(self):
        data = load_test_data()
        high = np.asarray(data["high"][:128], dtype=np.float64)
        low = np.asarray(data["low"][:128], dtype=np.float64)
        close = np.asarray(data["close"][:128], dtype=np.float64)
        result = mp.random_walk_index_batch(high, low, close, length_range=(14, 16, 2))
        assert result["rows"] == 2
        assert result["cols"] == high.shape[0]
        np.testing.assert_array_equal(result["lengths"], np.array([14, 16], dtype=np.uint64))

        single = mp.random_walk_index(high, low, close, length=14)
        np.testing.assert_allclose(result["high"][0], single["high"], rtol=0.0, atol=1e-12, equal_nan=True)
        np.testing.assert_allclose(result["low"][0], single["low"], rtol=0.0, atol=1e-12, equal_nan=True)

    def test_kernel_parity(self):
        data = load_test_data()
        high = np.asarray(data["high"][:144], dtype=np.float64)
        low = np.asarray(data["low"][:144], dtype=np.float64)
        close = np.asarray(data["close"][:144], dtype=np.float64)
        auto = mp.random_walk_index(high, low, close, length=14)
        scalar = mp.random_walk_index(high, low, close, length=14, kernel="scalar")
        np.testing.assert_allclose(auto["high"], scalar["high"], rtol=0.0, atol=1e-12, equal_nan=True)
        np.testing.assert_allclose(auto["low"], scalar["low"], rtol=0.0, atol=1e-12, equal_nan=True)

    def test_invalid_length(self):
        data = load_test_data()
        high = np.asarray(data["high"][:32], dtype=np.float64)
        low = np.asarray(data["low"][:32], dtype=np.float64)
        close = np.asarray(data["close"][:32], dtype=np.float64)
        with pytest.raises(ValueError, match="Invalid length"):
            mp.random_walk_index(high, low, close, length=0)

    def test_length_mismatch(self):
        data = load_test_data()
        high = np.asarray(data["high"][:32], dtype=np.float64)
        low = np.asarray(data["low"][:31], dtype=np.float64)
        close = np.asarray(data["close"][:32], dtype=np.float64)
        with pytest.raises(ValueError, match="Inconsistent slice lengths"):
            mp.random_walk_index(high, low, close, length=14)
